cmake/information_helpers.cmake

```diff
@@ -1,11 +1,104 @@
+    string(REGEX REPLACE "[$<A-Z_:]*SHELL:(.+)[>]*" "\\1" TMP "${TMP}")
+    string(REGEX REPLACE ".+INTERFACE:.+>" "" TMP "${TMP}")
+    string(REGEX REPLACE "\$<COMMA>" "," TMP "${TMP}")
+    set(${OUTPUT} "${TMP}" PARENT_SCOPE)
+endfunction()
+
+macro(ginkgo_interface_libraries_recursively INTERFACE_LIBS)
+    foreach(_libs ${INTERFACE_LIBS})
+        if (TARGET ${_libs})
+            if (upper_CMAKE_BUILD_TYPE STREQUAL "DEBUG" AND "${_libs}" MATCHES "ginkgo.*")
+                set(GINKGO_INTERFACE_LIB_NAME "-l${_libs}${CMAKE_DEBUG_POSTFIX}")
+            elseif("${_libs}" MATCHES "ginkgo.*") # Ginkgo libs are appended in the form -l
+                set(GINKGO_INTERFACE_LIB_NAME "-l${_libs}")
+            endif()
+
+            # Get the link flags and treat them
+            get_target_property(GINKGO_INTERFACE_LIBS_LINK_FLAGS "${_libs}"
+                INTERFACE_LINK_OPTIONS)
+            if (GINKGO_INTERFACE_LIBS_LINK_FLAGS)
+                filter_generator_expressions("${GINKGO_INTERFACE_LIBS_LINK_FLAGS}"
+                    GINKGO_INTERFACE_LIB_NAME)
+                list(APPEND GINKGO_INTERFACE_LIBS_FOUND "${GINKGO_INTERFACE_LIB_NAME}")
+            elseif (NOT "${GINKGO_INTERFACE_LIB_NAME}" IN_LIST GINKGO_INTERFACE_LIBS_FOUND)
+                list(APPEND GINKGO_INTERFACE_LIBS_FOUND "${GINKGO_INTERFACE_LIB_NAME}")
+            endif()
+
+            # Populate the include directories
+            get_target_property(GINKGO_LIBS_INTERFACE_INCS "${_libs}"
+                INTERFACE_INCLUDE_DIRECTORIES)
+            foreach(_incs ${GINKGO_LIBS_INTERFACE_INCS})
+                filter_generator_expressions("${_incs}" GINKGO_INTERFACE_INC_FILTERED)
+                if (GINKGO_INTERFACE_INC_FILTERED AND NOT
+                        "-I${GINKGO_INTERFACE_INC_FILTERED}" IN_LIST GINKGO_INTERFACE_CFLAGS_FOUND)
+                    list(APPEND GINKGO_INTERFACE_CFLAGS_FOUND "-I${GINKGO_INTERFACE_INC_FILTERED}")
+                endif()
+            endforeach()
+
+            # Populate the compiler options and definitions if needed
+            get_target_property(GINKGO_LIBS_INTERFACE_DEFS "${_libs}"
+                INTERFACE_COMPILE_DEFINITIONS)
+            if (GINKGO_LIBS_INTERFACE_DEFS)
+                list(APPEND GINKGO_INTERFACE_CFLAGS_FOUND "${GINKGO_LIBS_INTERFACE_DEFS}")
+            endif()
+            get_target_property(GINKGO_LIBS_INTERFACE_OPTS "${_libs}"
+                INTERFACE_COMPILE_OPTIONS)
+            filter_generator_expressions("${GINKGO_LIBS_INTERFACE_OPTS}" GINKGO_LIBS_INTERFACE_OPTS_FILTERED)
+            if (GINKGO_LIBS_INTERFACE_OPTS)
+                list(APPEND GINKGO_INTERFACE_CFLAGS_FOUND "${GINKGO_LIBS_INTERFACE_OPTS_FILTERED}")
+            endif()
+
+            # Keep recursing through the libraries
+            get_target_property(GINKGO_LIBS_INTERFACE_LIBS "${_libs}"
+                INTERFACE_LINK_LIBRARIES)
+            ginkgo_interface_libraries_recursively("${GINKGO_LIBS_INTERFACE_LIBS}")
```

Comment:
I think it will be a problem when ginkgo contain circular link? but we should not have this kind of issue because we always check the circular deps?

Reply:
when _libs is already in the list, it does not need to go recursively again.
for example, many ginkgo link ginkgo_device and ginkgo_device try to find hwloc